notifications/source/client.go

```diff
@@ -0,0 +1,187 @@
+package source
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"iter"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/icinga/icinga-go-library/notifications/event"
+	"github.com/pkg/errors"
+)
+
+// ErrRulesOutdated implies that the rules version between Icinga DB and Icinga Notifications mismatches.
+var ErrRulesOutdated = fmt.Errorf("rules version is outdated")
+
+// BasicAuthTransport is an http.RoundTripper that adds basic authentication and a User-Agent header to HTTP requests.
+type BasicAuthTransport struct {
+	http.RoundTripper // RoundTripper is the underlying HTTP transport to use for making requests.
+
+	Username   string
+	Password   string
+	ClientName string // ClientName is used to set the User-Agent header.
+}
+
+// RoundTrip adds basic authentication headers to the request and executes the HTTP request.
+func (b *BasicAuthTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.SetBasicAuth(b.Username, b.Password)
+	// As long as our round tripper is used for the client, the User-Agent header below
+	// overrides any other value set by the user.
+	req.Header.Set("User-Agent", b.ClientName)
+
+	return b.RoundTripper.RoundTrip(req)
+}
+
+// Client provides a common interface to interact with the Icinga Notifications API.
+// It holds the configuration for the API endpoint and the HTTP client used to make requests.
+type Client struct {
+	cfg Config // cfg holds base API endpoint URL and authentication details.
+
+	client http.Client // HTTP client used for making requests to the Icinga Notifications API.
+
+	IcingaWebBaseUrl *url.URL // IcingaWebBaseUrl holds the base URL for Icinga Web 2.
+	Endpoints        struct {
+		ProcessEvent string // ProcessEvent holds the URL for the process event endpoint.
+	}
+}
+
+// NewClient creates a new Client instance with the provided configuration.
+//
+// The projectName is used to set the User-Agent header in HTTP requests sent by this client and should be
+// set to the name of the project using this client (e.g., "Icinga DB").
+//
+// It may return an error if the API base URL or Icinga Web 2 base URL cannot be parsed.
+func NewClient(cfg Config, projectName string) (*Client, error) {
+	client := &Client{
+		cfg: cfg,
+		client: http.Client{
+			//Timeout: cfg.Timeout, // Uncomment once Timeout is (should be?) user configurable.
+			Transport: &BasicAuthTransport{
+				RoundTripper: http.DefaultTransport,
+				Username:     cfg.Username,
+				Password:     cfg.Password,
+				ClientName:   projectName,
+			},
+		},
+	}
+
+	baseUrl, err := url.Parse(cfg.ApiBaseUrl)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to parse API base URL")
+	}
+
+	client.Endpoints.ProcessEvent = baseUrl.ResolveReference(&url.URL{Path: "/process-event"}).String()
```

Comment:
Why does it use the more complex `ResolveReference()` over `JoinPath()`? This just makes it possible to have strange configuration like `ApiBaseUrl = "http://localhost:1234/what-am-i-even-doing"` and I don't see a good reason why that should work.

Reply:
Dunno. Changed.